File: frontend/src/lib/timerange.ts
/*
 * The console's time range — one vocabulary, one control, one window.
 *
 * Every ranged surface used to carry its own preset table: `MetricsChart` had
 * four windows, `LogExplorer` five, the audit trail six and an "All". So two
 * charts side by side were independently scoped and could honestly disagree
 * about what "now" covered, and the same words meant different spans depending
 * on which card you read them in.
 *
 * Two rules fix that and neither is negotiable. **The vocabulary is the one the
 * server already resolves** (`pkg/api/timerange.go`) rather than a second table
 * that agrees with it by coincidence. And **the browser never computes the
 * boundary**: a surface sends the preset's id and the server subtracts from its
 * own clock, so a chart, a trail and a link somebody pastes into a ticket all
 * name the same instant. That is also why these entries carry no `minutes`
 * field — there is nothing here to subtract with, on purpose.
 *
 * `all` is the one entry whose meaning is a property of the surface: the audit
 * trail reads it as "no lower bound", while a query against a datasource reads
 * it as the widest window that path allows, because a metrics backend has
 * retention and thirty days is the cap the query builder enforces anyway.
 */

export const TIME_RANGES = [
  { id: '15m', label: 'Last 15 minutes', short: '15m' },
  { id: '1h', label: 'Last hour', short: '1h' },
  { id: '6h', label: 'Last 6 hours', short: '6h' },
  { id: '24h', label: 'Last 24 hours', short: '24h' },
  { id: '7d', label: 'Last 7 days', short: '7d' },
  { id: '30d', label: 'Last 30 days', short: '30d' },
  { id: 'all', label: 'All time', short: 'All' },
] as const

export type TimeRangeId = (typeof TIME_RANGES)[number]['id']

/**
 * The window the console opens on. A day rather than an hour, because the
 * surface most damaged by the wrong default is the audit trail: a chart over a
 * day is coarser than one over an hour, while a trail over an hour is usually
 * empty, and an empty page reads as a broken one.
 */
export const DEFAULT_TIME_RANGE: TimeRangeId = '24h'

/** The query parameter the window travels in, so a link carries what it was read at. */
export const TIME_RANGE_PARAM = 'range'

export function isTimeRange(value: string | null | undefined): value is TimeRangeId {
  return TIME_RANGES.some((entry) => entry.id === value)
}

export function timeRangeLabel(id: TimeRangeId): string {
  return TIME_RANGES.find((entry) => entry.id === id)?.label ?? id
}

/**
 * The same window, named for a surface that reads a datasource. `all` is the
 * only entry that differs: the query path caps a single query at thirty days,
 * so a chart claiming to draw all time would be claiming more than it drew.
 */
export function queryRangeLabel(id: TimeRangeId): string {
  return id === 'all' ? 'Last 30 days (the widest a query allows)' : timeRangeLabel(id)
}

File: backend/pkg/api/timerange.go
package api

import (
	"net/http"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
)

/*
 * The console's time range.
 *
 * "The last hour" has to mean the same span in a chart, in the audit trail and
 * in a link somebody pastes into a ticket. None of those agree if the browser
 * computes the boundary — three surfaces each subtracting from their own clock
 * produce three windows, and a row count that disagrees with the page it counts
 * is a trail nobody trusts. So the vocabulary is a fixed table here and the
 * boundary is resolved server-side, from this process's clock, for every surface
 * that takes a range.
 *
 * It is a table rather than a parsed duration for the same reason the audit
 * filter has always used one: the set the console offers and the set the API
 * accepts cannot drift apart, and no caller can ask for a window wide enough to
 * be a table scan dressed as a filter.
 */

// consoleRanges are the presets every ranged surface shares.
var consoleRanges = map[string]time.Duration{
	"15m": 15 * time.Minute,
	"1h":  time.Hour,
	"6h":  6 * time.Hour,
	"24h": 24 * time.Hour,
	"7d":  7 * 24 * time.Hour,
	"30d": 30 * 24 * time.Hour,
}

// rangeSpan resolves the `range` parameter into a span back from now.
//
// `all` is the one preset whose meaning is a property of the surface rather than
// of the vocabulary, so the caller supplies it: for the audit trail it is "no
// lower bound at all", and for a query against a datasource it is the widest
// window that path allows — a metrics backend has retention, and asking for
// everything it holds is asking for as far back as the query is permitted to
// reach. Zero means "no preset", which is also what an absent parameter gives.
func rangeSpan(c *gin.Context, all time.Duration) (time.Duration, bool) {
	raw := strings.ToLower(strings.TrimSpace(c.Query("range")))
	if raw == "" {
		return 0, true
	}
	if raw == "all" {
		return all, true
	}
	window, known := consoleRanges[raw]
	if !known {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": "range must be one of 15m, 1h, 6h, 24h, 7d, 30d or all",
		})
		return 0, false
	}
	return window, true
}
